View review decision

Slartibartfass2 opened the pull request:
Closes #45 

## What I have made

- added lists with criteria and review info
- added the final paper decision banner below

Use the following files to test it without the mock backend:
[mock-data.zip](https://github.com/user-attachments/files/18871599/mock-data.zip)
- put `+layout.ts` in `src/routes/project/[projectId]`
- put `+page.ts` in `src/routes/project/[projectId]/paper/[paperId]`

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - ~~[ ] unit tests~~ (no logic added that needs unit tests)
  - [x] integration tests
  - ~~[ ] end-to-end tests~~ (not yet available)
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 committed:
feat: add separator component

Slartibartfass2 committed:
feat: add checkbox component

Slartibartfass2 committed:
feat: add CriterionListEntry component

Slartibartfass2 committed:
feat: add CriterionListEntrySkeleton component

Slartibartfass2 committed:
refactor: extract promise wrapper for model builder

This way, each component can be wrappen in `loading` to make it async.

Slartibartfass2 committed:
test: add tests for CriterionListEntry

Slartibartfass2 committed:
feat: add CriteriaList and ReviewCriteriaList component

Slartibartfass2 committed:
feat: add review criteria to paper view

Slartibartfass2 committed:
feat: add PaperDecisionBanner component

Merseleo commented on src/lib/components/composites/Tooltip.svelte:
```diff
@@ -8,17 +12,18 @@
+    const {
+        trigger,
+        content,
+        triggerVariant = "none",
+        triggerSize = "fit",
```
What happens in this default case, because I deleted the "triggerVariant" and "triggerSize" in the `PaperBookmarkButton` (so the defaults should be used), but nothing happened?!

Slartibartfass2 replied:
Good catch. Yes, indeed, we set the whole style with the class attribute, so these attributes don't make a difference for the `PaperBookmarkButton` component.

Merseleo replied:
Ok.

Merseleo commented on src/lib/components/composites/user-avatar/UserAvatarSkeleton.svelte:
Great idea to outsource this Skeleton. Maybe you can also use it in the navigation bar?

Slartibartfass2 replied:
Where? I can't find the place where a user avatar skeleton is used.

Merseleo replied:
Currently there is no user avatar skeleton used there. I thought we could add it, when the current user is loaded and pass a `Promise<User>` to the NavigationBar instead of the `User`. But the user is only loaded once, when logging in and then everything is loaded, so my idea doesn't make sense anyway. Forget about it.

Merseleo commented on src/lib/components/composites/user-avatar/UserAvatar.svelte:
````diff
@@ -32,15 +32,15 @@ in the PaperEntry component.
 Usage:
 ```svelte
-    <UserAvatar userId={{
+    <UserAvatar user={{
````
Great catch, thank you ^^.

Slartibartfass2 replied:
:)

Merseleo commented on src/lib/components/composites/criteria/CriteriaList.svelte:
```diff
@@ -0,0 +1,57 @@
+    import CriterionListEntrySkeleton from "./CriterionListEntrySkeleton.svelte";
+
+    interface Props {
+        listTitle: string;
```
```suggestion
        listTitle: "Hard Exclusion" | "Soft Exclusion" | "Inclusion";
```
as we actually want only this three types of criteria lists and it should not be easily possible to add a new list type, as this would require a lot of more work, dont you agree?

Slartibartfass2 replied:
Good catch. I changed it.

Merseleo commented on src/lib/components/composites/criteria/CriteriaList.svelte:
```diff
@@ -0,0 +1,57 @@
+
+<!--
+@component
+A list of criteria.
```
Maybe you could extend this comment on what is shown depending on the review mode?

Slartibartfass2 replied:
I referred to the documentation of `CriteriaListEntry` where this is explained already.

Merseleo replied:
Good idea, thank you.

Merseleo commented on src/lib/components/composites/criteria/CriteriaList.svelte:
````diff
@@ -0,0 +1,57 @@
+```
+-->
+<section class="flex flex-col gap-5">
+    <h2>{listTitle}</h2>
````
haha, today in the exam I just learned to use the correct semantic tags, so h1 and style via css. We just broke that ...

Slartibartfass2 replied:
I don't get it. What's wrong with that?

Merseleo commented on src/lib/components/composites/criteria/CriterionListEntrySkeleton.svelte:
```diff
@@ -0,0 +1,27 @@
+    {#if inReviewMode}
+        <div class="size-4.5"></div>
+    {/if}
+    <div class="flex flex-row gap-1.5 py-0.5">
```
```suggestion
    <div class="flex flex-row gap-2.5 py-0.5">
```

Slartibartfass2 replied:
I made the same changes as in `CriterionListEntry`. Let's discuss it there.

Merseleo replied:
I like your solution with the gap of 8px.

Merseleo commented on src/lib/components/composites/criteria/CriteriaList.svelte:
```diff
@@ -0,0 +1,57 @@
+            {/if}
+        {:catch error}
+            {console.error(`Failed to load criteria: ${error}`)}
+            <span class="text-error">Couldn't load Criteria</span>
```
```suggestion
            <div class="flex flex-row items-center gap-x-2 p-4">
            <CircleAlert class="text-neutral-500" size={20} />
            <span class="text-error">Couldn't load criteria</span>
        </div>
```

Slartibartfass2 replied:
We use this fairly regularly. I extracted it to a component.

Merseleo replied:
Ok, thank you.

Merseleo commented on src/lib/components/composites/criteria/CriterionListEntry.svelte:
```diff
@@ -0,0 +1,64 @@
+
+<!--
+@component
+Single Criterion List Element.
```
```suggestion
@component
Single list element for a criterion.
```

Slartibartfass2 replied:
Changed it.

Merseleo commented on src/lib/components/composites/criteria/CriterionListEntry.svelte:
````diff
@@ -0,0 +1,64 @@
+    </ul>
+```
+-->
+<li class="flex flex-row gap-2.5 items-center" data-testid="criterion-list-entry">
````
A distance of 10 pixels makes the individual components look so squashed, especially the user avatar, and a little more distance would perhaps be helpful in general. I also thought we had it differently in the mockup.

Merseleo replied:
But I remember correctly the gap between the tag and name is 2.5

Slartibartfass2 replied:
The gaps have the same value as or a higher value than the mockup. I increased them. What do you think of it now?
I also added truncation if the name is too long.

Merseleo replied:
Good idea with the truncation, didnt consider this. I like the gap of 16px, however imo we should group the user avatars in a div and decrease the gap for them, what do you think?

Slartibartfass2 replied:
I grouped them together. What is your opinion now?

Merseleo replied:
Looks better, I think we should leave it like this.

Merseleo commented on src/lib/components/composites/criteria/CriteriaList.svelte:
```diff
@@ -0,0 +1,57 @@
+-->
+<section class="flex flex-col gap-5">
+    <h2>{listTitle}</h2>
+    <ul class="flex flex-col gap-4 pl-2">
```
Regarding semantic tags: actually this is a \<dl\> isn't it?

Slartibartfass2 replied:
I read the documentation of `<dl>` and it doesn't fit in my opinion. Why do you think it should be an `<dl>` tag?
I chose `<ul>` because it's an unordered list of criteria.

Merseleo replied:
I thought the tag could be interpreted as term and the remaining parts as description, but `<ul>` is also correct. Lerts keep it as it is.

Merseleo commented on src/lib/components/composites/criteria/CriterionListEntry.svelte:
```diff
@@ -0,0 +1,64 @@
+    </Tooltip>
+    {#if !inReviewMode}
+        {#each criterion.reviews as review (review.id)}
+            <UserAvatar
```
Do you think it is possible to make the user avatar smaller here? Because it is quite big compared to the text (same in the decision banner)? What do you think of a smaller user avatar?

Slartibartfass2 replied:
I made a smaller avatar, which is about 80% of the original size.

Merseleo replied:
Imo, it looks better now. Thank you.

Merseleo commented on src/lib/components/composites/criteria/PaperDecisionBanner.svelte:
```diff
@@ -0,0 +1,79 @@
+            case PaperDecision.DECLINED:
+                return "bg-decline-red";
+            default:
+                return "bg-unreviewed-gray";
```
I think this is not completely correct, as you also return unreviewed-gray when the decision is "Maybe".

Merseleo replied:
Same for the decision label. Maybe undecided could be a bit too unspecific?

Slartibartfass2 replied:
A `PaperDecision` can't be "maybe". Only review decisions.

Slartibartfass2 replied:
The default case is for `UNDECIDED` and `UNSPECIFIED`

Merseleo replied:
I know that a `PaperDecision` can't be "maybe", but you can figure out, whether the undecided is "not reviewed undecided" or ("maybe undecided") by additionally checking, whether there are any reviews.
And if there are, maybe you could return "bg-maybe-yellow" or so.

Slartibartfass2 replied:
As discussed, we show "Unreviewed", when no reviews are available, and otherwise we show "Undecided" with the maybe color.

Merseleo replied:
Thank you, I tested it with all four "decisions" and it seems more user-friendly.

Merseleo commented on src/lib/components/composites/criteria/PaperDecisionBanner.svelte:
```diff
@@ -0,0 +1,79 @@
+    const bannerLabel = resource<Project_Paper, string>(loadingProjectPaper, {
+        initialValue: "",
+        onSuccess: (paper) => getDecisionLabel(paper.decision),
+        onErrorValue: "Couldn't load Paper Decision",
```
```suggestion
        onErrorValue: "Couldn't load paper decision",
```

This is a general point we could discuss. You always use uppercase letters at the beginning, but I thought in englisch everything is written in lowercase?!

Slartibartfass2 replied:
Yes, I thought this looks better and because it's a specific term in our application, I wrote it in upper case. It doesn't really bother me, how we write it.

Merseleo replied:
But do we really call it "Paper Decision" somewhere? Imo it is better to dont write in upper case or at least dont mix it.

Slartibartfass2 replied:
Yes, I agree.

Merseleo commented on src/lib/components/composites/criteria/PaperDecisionBanner.svelte:
```diff
@@ -0,0 +1,79 @@
+        <div class="flex flex-row gap-2">
+            <UserAvatarSkeleton />
+            <UserAvatarSkeleton />
+        </div>
```
Why no extra skeleton component here?

Slartibartfass2 replied:
Didn't see a reason why. Do you think we should extract a component there?

Merseleo replied:
Not necessarily, it is quite small, only wanted to check, whether there was a specific reason or not.

Merseleo commented on src/lib/components/composites/criteria/PaperDecisionBanner.svelte:
```diff
@@ -0,0 +1,79 @@
+        <h2 class={projectPaper.decision === PaperDecision.ACCEPTED ? "text-black" : "text-white"}>
+            {bannerLabel.value}
+        </h2>
+        <div class="flex flex-row gap-2">
```
```suggestion
        <div class="flex flex-row gap-3">
```

Slartibartfass2 replied:
Changed it.

Merseleo commented on src/lib/components/composites/criteria/PaperDecisionBanner.svelte:
```diff
@@ -0,0 +1,79 @@
+                <UserAvatar
+                    reviewDecision={review.decision}
+                    user={reviewers.find((reviewer) => review.userId === reviewer.id)}
+                />
```
Same here, maybe a smaller user avatar could look better, what do you think?

Slartibartfass2 replied:
Already changed it.

Merseleo commented on src/lib/components/composites/criteria/PaperDecisionBanner.svelte:
```diff
@@ -0,0 +1,79 @@
+        </div>
+    {:catch error}
+        {console.error(`Failed to load paper decision: ${error}`)}
+        <span class="text-error text-white">Couldn't load Paper Decision</span>
```
```suggestion
        <div class="flex flex-row items-center gap-x-2 p-4">
            <CircleAlert class="text-neutral-500" size={20} />
            <span class="text-error">Couldn't load paper decision</span>
        </div>
```

Slartibartfass2 replied:
Used the new component here as well.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/PaperView.svelte:
```diff
@@ -83,11 +102,17 @@ Usage:
-            {forwardReferencedPapers}
-            inReviewMode={userConfig.isReviewMode}
-        />
+        <!-- TODO: figure out what to show when it's the project independent paper view -->
```
Maybe display the `PaperResearchContextCard` but without the second tab for "Review Information"?

Slartibartfass2 replied:
Working on it...

Merseleo replied:
I checked your solution and it looks great. Seems more intuitive now, that the normal paper view doesn't have a tab for review information.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/PaperView.svelte:
```diff
@@ -98,8 +123,14 @@ Usage:
                     <DeclineButton {loadingPaperId} />
-                    {#if userConfig.showMaybeButton}
-                        <MaybeButton {loadingPaperId} />
+                    {#if loadingProject}
+                        {#await loadingProject}
+                            <!-- show nothing while loading -->
+                        {:then project}
+                            {#if project.settings?.reviewMaybeAllowed}
+                                <MaybeButton {loadingPaperId} />
+                            {/if}
+                        {/await}
                     {/if}
```
I think we should continue with requesting these settings (as well as the settings for the review mode) at a higher level and save these settings in a store we can always access. This would make a lot of code more simple.

Slartibartfass2 replied:
I agree, we should store the user config somewhere.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/PaperResearchContextCard.svelte:
```diff
@@ -34,6 +48,8 @@ Usage:
     <PaperCardContent value="1">
         {#if inReviewMode}
+            <ReviewCriteriaList {inReviewMode} {reviewedCriteria} {reviewers} />
+            <Separator />
```
What do you think of styling this separator here? Because it is very thin and barely visible.

Slartibartfass2 replied:
Looked fine on my machine. How thick should it be in your opinion?

Merseleo replied:
Something like a height of 2 or 3 pixels instead of 1, but actually this a component which is always thin. So if I rethink my suggestion: lets keep it as it is.

Merseleo commented on src/lib/components/composites/paper-components/paper-view/cards/ReviewCriteriaList.svelte:
```diff
@@ -0,0 +1,50 @@
+    {inReviewMode}
+    listTitle="Inclusion"
+    {reviewers}
+/>
```
Maybe you can wrap this with a div with a spacing of 20px, like in the Figma to visually separate these lists. What do you think of this idea?

Slartibartfass2 replied:
I totally forgot that. I added it now.

Merseleo commented on src/lib/model/general.ts:
```diff
@@ -46,5 +48,23 @@ function asPaper(paper: Project_Paper | Paper): Paper {
+    reviews: Omit<Review, "selectedCriteriaIds">[];
+};
+
+interface UserConfig {
+    isReviewMode: boolean;
+}
```
Does it make sense to add the showMaybeButton setting too?

Slartibartfass2 replied:
the maybe button setting is stored in the project settings, not the user settings.

Merseleo replied:
Sorry, my mistake.

Merseleo commented on src/routes/project/[projectId]/paper/[paperId]/+page.ts:
I think it is better to finally review this later, as I do not really know how to test this with all the problems I currently have with the mock backend, but it seems like it should work properly!

Slartibartfass2 replied:
Yes, this sucks a lot!

Merseleo commented on src/routes/project/[projectId]/paper/[paperId]/+page.ts:
```diff
@@ -21,10 +25,70 @@ export const load: PageLoad = ({ params }) => {
+ * @param reviews - List of reviews
+ * @returns List of reviewed criteria
+ */
+async function createReviewedCriteria(
```
The comment helped me, but I struggled a bit to understand, what a reviewedCriterion is. Do have ideas for a better name? Or it is only me?

Slartibartfass2 replied:
I'm a bit clueless on how I should name that.

These are the proposals of ChatGPT:
1. CriterionWithReviews
2. ReviewedCriterion
3. CriterionWithAssignments
4. FullCriterion
5. CriterionWithFeedback
6. CriterionDetails
7. CriterionWithReviewsData
8. CriterionReviewBundle
9. CompleteCriterion
10. CriterionReviewSet

Merseleo replied:
I think the name "CriterionWithReviews" is a good option.

Slartibartfass2 replied:
Alright, I changed it.

Merseleo replied:
Thank you, I read some code and comments again and now it makes more sense.

Merseleo commented on src/routes/project/[projectId]/paper/[paperId]/+page.ts:
```diff
@@ -21,10 +25,70 @@ export const load: PageLoad = ({ params }) => {
+        const filteredReviews = reviews.filter((review) =>
+            review.selectedCriteriaIds.includes(criterion.id),
+        );
+        for (const review of filteredReviews) {
+            criterionReviews.push({
+                id: review.id,
+                decision: review.decision,
+                userId: review.userId,
+            });
+        }
```
Only a question of interest: Why dont you use the foreach directly after the filter?

Slartibartfass2 replied:
Not sure, tbh. Maybe I had a different approach at the beginning and required a for loop.

Merseleo replied:
Ok.

Merseleo commented on tests/integration/criteria/criteria-list.test.ts:
```diff
@@ -0,0 +1,80 @@
+        expect(hint).toBeInTheDocument();
+    });
+
+    test("When criteria failed to load, then error message is shown", async () => {
```
```suggestion
    test("When no criterion could be loaded, then the error message is shown", async () => {
```

Slartibartfass2 replied:
Successfully loading no criteria is not the same as failing to load criteria.

Merseleo replied:
That's why "When no criterion could be loaded" instead of "When no criterion is loaded", but I understand your point. Probably it would be confusing.

Merseleo commented on tests/integration/criteria/criterion-list-entry.test.ts:
```diff
@@ -0,0 +1,68 @@
+        expect(name).toBeInTheDocument();
+    });
+
+    test("When `isReviewMode` is true, then the checkbox is shown", () => {
```
```suggestion
    test("When `inReviewMode` is true, then the checkbox but no user avatars are shown", () => {
```

In gerenal, I like the idea to test both things in one test, but even if the reviewMode would be true, your last test case would pass, as you dont add any reviewers that could be shown, but aren't due to the `inReviewMode` flag.

Slartibartfass2 replied:
True that. I added them.

Merseleo commented on tests/integration/criteria/criterion-list-entry.test.ts:
```diff
@@ -0,0 +1,68 @@
+        expect(userAvatar).not.toBeInTheDocument();
+    });
+
+    test("When `isReviewMode` is false, then the reviews are shown", () => {
```
```suggestion
    test("When `inReviewMode` is false, then the reviews are shown", () => {
```

Slartibartfass2 replied:
Good catch!

Merseleo commented on src/lib/components/composites/criteria/CriterionListEntry.svelte:
```diff
@@ -0,0 +1,64 @@
+        <span class="font-bold">{criterion.tag}</span>
+        <span>{criterion.name}</span>
+    </div>
+    <Tooltip
```
What do you think of the idea of making this a button and not a tooltip, so it is not only hoverable but also clickable?

Slartibartfass2 replied:
Why should it be clickable?

Merseleo replied:
Because I was acted as a naive user (because I am a naive and dumb user) and already tried to click on the help icon and nothing happened. But it was only an idea.

Slartibartfass2 replied:
I agree that this would increase the usability. I added it.

Merseleo replied:
Thank you.

Merseleo commented on tests/integration/criteria/paper-decision-banner.test.ts:
I think you could a test for checking, that the user avatars with their individual decisions are shown?!

Slartibartfass2 replied:
Good catch. I added it.

Merseleo commented on tests/integration/paper-components/paper-view/cards/review-criteria-list.test.ts:
```diff
@@ -0,0 +1,170 @@
+describe("ReviewCriteriaList", () => {
+    test("When props are provided, then component is shown", () => {
+        render(ReviewCriteriaList, {
+            target: document.body,
```
I have already noticed this in some previous tests: Why do you explicitly define the target? Isn't it automatically document.body?

Slartibartfass2 replied:
I started it at some point and now I'm keeping it consistent 😅

Merseleo replied:
Ok

Merseleo commented on tests/integration/paper-components/paper-view/cards/review-criteria-list.test.ts:
```diff
@@ -0,0 +1,170 @@
+        const softExclusions = screen.getByText("Soft Exclusion");
+        expect(softExclusions).toBeInTheDocument();
+        const inclusions = screen.getByText("Inclusion");
+        expect(inclusions).toBeInTheDocument();
```
Shouldnt you also expect a decision banner?

Slartibartfass2 replied:
No, the decision banner is part of the page, not the `CriteriaList` component.

Merseleo replied:
You are right, my mistake.

Merseleo commented on tests/model-builder.ts:
I am beginning to like this model builder file ...

Slartibartfass2 replied:
:)))

Merseleo reviewed (changes requested):
There are a lot of comments, but many of them are ideas or spelling mistakes or something like that. In general, it looks good, even if it deviates from the mockup in terms of design and looks a bit squashed. But still good work! I am looking forward to your responses regarding some suggestions or questions.

And one general idea, that came to my mind: what do you think of displaying the full name or user name when hovering over the user avatar?

Slartibartfass2 committed:
refactor: extract error indicator to separate component

Slartibartfass2 committed:
feat: add new size 'small' for user avatar

Slartibartfass2 reviewed (commented):
Thank you for taking the time to review this. I'll address the issue with displaying the right card in the project independent paper view ASAP. Meanwhile, I'll push all my changes.

For the hoverable user avatar, we already have https://github.com/SE-UUlm/snowballr-frontend/issues/123.

Slartibartfass2 committed:
chore: address review

mostly minor style changes

Slartibartfass2 committed:
refactor: make research context card easier to switch.

Now, only change the tabs but the code for the content keeps the same. This removes the if blocks.

Slartibartfass2 committed:
feat: show only references tab in non-project paper view

Slartibartfass2 commented:
We now display only the references tab, when the non-project paper view is shown. It was a pain to get the types right, and I'm not happy with the result. Maybe, you find a different way so that the types of the props can be inferred.

Merseleo commented on src/lib/components/composites/criteria/PaperDecisionBanner.svelte:
```diff
@@ -51,7 +51,7 @@
 <div
     class={cn(
-        "flex flex-row w-full rounded-2xl items-center justify-center p-2.5 gap-5",
+        "flex flex-col lg:flex-row w-full rounded-2xl items-center justify-center p-2.5 gap-3 lg:gap-5",
```
Great idea, but consider to first decrease the "margin" of the banner, because as you can see here:
<img width="416" alt="Bildschirmfoto 2025-02-24 um 14 49 05" src="https://github.com/user-attachments/assets/7416a45b-4c54-4f53-8c63-c98ee0bfa2b7" />
I think there is enough space left to keep a row layout, if the banner has more space.

Slartibartfass2 replied:
I changed the behavior, so that more space is taken, and the content is wrapped.

Merseleo replied:
I tested it and I like the new "shrinking" behavior way more!

Merseleo commented on src/lib/components/composites/paper-components/paper-view/PaperView.svelte:
```diff
@@ -57,6 +60,25 @@
+            reviewedCriteria,
+            loadingProjectPaper: loadingPaperWrapper,
+        };
+    } else {
```
Isnt the else block the same as the if block?

Slartibartfass2 replied:
Optically, yes, but this is the only way I got the type inference to work.

Merseleo replied:
Ok, then we keep it (but maybe you can explain this in a short comment)

Slartibartfass2 replied:
I added more comments.

Merseleo replied:
Perfect.

Merseleo commented:
> We now display only the references tab, when the non-project paper view is shown. It was a pain to get the types right, and I'm not happy with the result. Maybe, you find a different way so that the types of the props can be inferred.

In my opinion your solution may not be optimal, as the types are not inferred, but you can be happy with the result. I dont have a idea to do it better and I also think it is good enough.

Merseleo reviewed (changes requested):
I think we / you are nearly done with it, it looks great now.

Slartibartfass2 committed:
feat: add options to open a tooltip when being clicked

Slartibartfass2 committed:
refactor: rename ReviewedCriterion to CriterionWithReviews

Slartibartfass2 commented:
> > We now display only the references tab, when the non-project paper view is shown. It was a pain to get the types right, and I'm not happy with the result. Maybe, you find a different way so that the types of the props can be inferred.
> 
> In my opinion your solution may not be optimal, as the types are not inferred, but you can be happy with the result. I dont have a idea to do it better and I also think it is good enough.

Actually, the types are inferred now. This was important to me and I can accept a bit ugly code when I have better type-safety instead.

Slartibartfass2 reviewed (commented):
Thank you for the review.

Slartibartfass2 committed:
feat: minor style changes

Merseleo commented on tests/integration/criteria/paper-decision-banner.test.ts:
```diff
@@ -0,0 +1,162 @@
+        expect(label).toBeInTheDocument();
+    });
+
+    test("When the paper is undecided and has no reviews, it should display the undecided banner", async () => {
```
```suggestion
    test("When the paper is undecided and has no reviews, it should display the unreviewed banner", async () => {
```

Slartibartfass2 replied:
Oops, that's the result of copying and pasting stuff. I fixed it.

Merseleo reviewed (changes requested):
Only two minor things left and dont forget to rebase.

Slartibartfass2 reviewed (commented):
Let's finish this.
